File: Medium/61_Rotate_List/Solution.java



public class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null || k == 0) {
            return head;
        }

        int len = 0;
        ListNode curr = head;

        while (curr != null) {
            curr = curr.next;
            len++;
        }

        int rotate = k % len;

        if (rotate == 0){
            return head;
        }
        
        ListNode newTail = head;
        for (int i = 1; i < len - rotate; i++) {
            newTail = newTail.next;
        }

        ListNode newHead = newTail.next;
        newTail.next = null;
        ListNode temp = newHead;

        while (temp.next != null) {
            temp = temp.next;
        }
        temp.next = head;

        return newHead;
    }

    void printSinglyLinkList(ListNode head) {
        ListNode temp = head;
        while (temp != null) {
            System.out.print(temp.val + " ");
            temp = temp.next;
        }
        System.out.println();
    }
    
    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        ListNode head1 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
        int k1 = 2;
        ListNode result1 = solution.rotateRight(head1, k1);
        solution.printSinglyLinkList(result1); // Expected output: 4 -> 5 -> 1 -> 2 -> 3

        // test cases 2
        ListNode head2 = new ListNode(0, new ListNode(1, new ListNode(2)));
        int k2 = 4;
        ListNode result2 = solution.rotateRight(head2, k2);
        solution.printSinglyLinkList(result2); // Expected output: 2 -> 0 -> 1
    }
}

// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

File: Medium/61_Rotate_List/Solution.cpp
#include "../../devlibs/cpp/cpphelper.h"

using namespace std;


// NOTE:
// For LeetCode submission, copy only the `class Solution` part.


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
*/

class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head || !head->next || k == 0) return head;

        int len = 1;
        ListNode* curr = head;

        while (curr->next) {
            curr = curr->next;
            len++;
        }

        int rotate = k % len;

        if (rotate == 0) return head;

        ListNode* slow = head;
        ListNode* fast = head;

        for (int i = 0; i < rotate; i++) {
            fast = fast->next;
        }

        while (fast->next) {
            slow = slow->next;
            fast = fast->next;
        }

        ListNode* newHead = slow->next;
        slow->next = nullptr;
        fast->next = head;

        return newHead;
    }
};


int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    Solution solution;
    
    // test cases 1
    ListNode* head1 = createSinglyLinkList({1,2,3,4,5});
    int k1 = 2;
    ListNode* result1 = solution.rotateRight(head1, k1);
    printSinglyLinkList(result1); // Expected output: 4 -> 5 -> 1 -> 2 -> 3

    // test cases 1
    ListNode* head2 = createSinglyLinkList({0,1,2});
    int k2 = 4;
    ListNode* result2 = solution.rotateRight(head2, k2);
    printSinglyLinkList(result2); // Expected output: 2 -> 0 -> 1

    return 0;
}